clients/elrond/client.go

```diff
@@ -176,332 +196,124 @@ func (c *client) createPendingBatchFromResponse(responseData [][]byte) (*clients
-//	}
-//
-//	c.log.Debug("Elrond: get erc20 address", "tokenID", tokenId, "address", address)
-//
-//	return address
-//}
-//
-//// GetHexWalletAddress returns the wallet address as a hex string
-//func (c *client) GetHexWalletAddress() string {
-//	return hex.EncodeToString(c.address.AddressBytes())
-//}
+// ProposeSetStatus will trigger the proposal of the ESDT safe set current transaction batch status operation
+func (c *client) ProposeSetStatus(ctx context.Context, batch *clients.TransferBatch) (string, error) {
+	if batch == nil {
+		return "", errNilBatch
+	}
+
+	txBuilder := builders.NewTxDataBuilder().Function(proposeSetStatusFuncName).ArgInt64(int64(batch.ID))
+	txBuilder.ArgBytes(batch.Statuses)
+
+	hash, err := c.sendTransactionReturnHash(ctx, txBuilder, c.gasMapConfig.ProposeStatus)
+	if err == nil {
+		c.log.Info("proposed set statuses"+batch.String(), "transaction hash", hash)
+	}
+
+	return hash, err
+}
+
+// ResolveNewDeposits will try to add new statuses if the pending batch gets modified
+func (c *client) ResolveNewDeposits(ctx context.Context, batch *clients.TransferBatch) error {
+	if batch == nil {
+		return errNilBatch
+	}
+
+	newBatch, err := c.GetPending(ctx)
+	if err != nil {
+		return fmt.Errorf("%w while getting new batch in ResolveNewDeposits method", err)
+	}
+
+	batch.ResolveNewDeposits(len(newBatch.Statuses))
+
+	return nil
+}
+
+// ProposeTransfer will trigger the propose transfer operation
+func (c *client) ProposeTransfer(ctx context.Context, batch *clients.TransferBatch) (string, error) {
+	if batch == nil {
+		return "", errNilBatch
+	}
+
+	txBuilder := builders.NewTxDataBuilder().Function(proposeTransferFuncName).ArgInt64(int64(batch.ID))
+
+	for _, dt := range batch.Deposits {
+		txBuilder.ArgBytes(dt.FromBytes).
+			ArgBytes(dt.ToBytes).
+			ArgBytes(dt.ConvertedTokenBytes).
+			ArgBigInt(dt.Amount).
+			ArgInt64(int64(dt.Nonce))
+	}
+
+	gasLimit := c.gasMapConfig.ProposeTransferBase + uint64(len(batch.Deposits))*c.gasMapConfig.ProposeTransferForEach
+	hash, err := c.sendTransactionReturnHash(ctx, txBuilder, gasLimit)
+	if err == nil {
+		c.log.Info("proposed transfer"+batch.String(), "transaction hash", hash)
+	}
+
+	return hash, err
+}
+
+// Sign will trigger the execution of a sign operation
+func (c *client) Sign(ctx context.Context, actionID uint64) (string, error) {
+	txBuilder := builders.NewTxDataBuilder().Function(signFuncName).ArgInt64(int64(actionID))
+
+	hash, err := c.sendTransactionReturnHash(ctx, txBuilder, c.gasMapConfig.Sign)
+	if err == nil {
+		c.log.Info("signed", "action ID", actionID, "transaction hash", hash)
+	}
+
+	return hash, err
+}
+
+// PerformAction will trigger the execution of the provided action ID
+func (c *client) PerformAction(ctx context.Context, actionID uint64, batch *clients.TransferBatch) (string, error) {
+	if batch == nil {
+		return "", errNilBatch
+	}
+
+	txBuilder := builders.NewTxDataBuilder().Function(performActionFuncName).ArgInt64(int64(actionID))
```

Comment:
maybe this can be a new method, similar to createDefaultVmQueryBuilder:
```golang
func (c *client) createTxDataBuilder(funcName string, id int64) builders.TxDataBuilder {
	return builders.NewTxDataBuilder().Function(funcName).ArgInt64(id)
}
```

Reply:
done